File: KMP/28. Implement strStr()/Solution.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        N,M=len(haystack),len(needle)
        if M==0:
            return 0
        def buildTable(pattern):
            n=len(pattern)
            table=[0]*n
            i,j=1,0
            while i<n:
                if pattern[i]==pattern[j]:
                    table[i]=j+1
                    j+=1
                    i+=1
                else:
                    if j>0:
                        j=table[j-1]
                        # go back to the previous pattern
                        # find the matches
                        # keep in mind there is NO i++ here
                    else:
                        #just set table[i] to 0
                        table[i]=j # or table[i]=0
                        i+=1
            #shift table for convenice
            return [-1]+table[:-1]
        table=buildTable(needle)
        i,j=0,0
        while i<N and j<M:
            if haystack[i]==needle[j]:
                i+=1
                j+=1
            else:
                #re-try previous/other pattern
                j=table[j]
                #if no pattern left
                if j==-1:
                    # cannot find haystack[i] in needle[j]
                    # skip haystack[i], retry everything
                    i+=1
                    j=0 #or j+=1
        return (i-M) if j==M else -1

File: KMP/28. Implement strStr()/Solution.java
class Solution {
    public int strStr(String haystack, String needle) {
        int n=haystack.length();
        int m=needle.length();
        if (m==0){
            return 0;
        }
        int[] table=prefixTable(needle);//table shifted by 1
        int i=0, j=0;
        while (i<n&&j<m){
            if (haystack.charAt(i)==needle.charAt(j)){
                if (j==m-1)
                    return i-m+1;
                i++;
                j++;
            }else{
                j=table[j];
                if (j==-1){
                    i++;
                    j=0;//or j++
                }
            }
        }
        return (j==m)?(i-m+1):-1;
    }
    private int[] constructTable(String pattern){
        int m=pattern.length();
        int[] table=new int[m];//this is the table indicate how many prefix is matched
        for (int i=1, j=0; i<m; i++){
            while (j>0 && pattern.charAt(i)!=pattern.charAt(j)){
                j=table[j-1];
            }
            if (pattern.charAt(i)==pattern.charAt(j)){
                table[i]=j+1;
                j++;
            }else{
                table[i]=j;
            }
        }
        return table;
    }
    private int[] prefixTable(String pattern){
        int m=pattern.length();
        int i=1, j=0;
        int[] table=new int[m];
        while (i<m){
            if (pattern.charAt(i)==pattern.charAt(j)){
                j++;
                table[i++]=j;
            }else{
                if (j>0){
                    j=table[j-1];
                }else{
                    table[i++]=j; // that just mean table[i]=0
                }
            }
        }
        for (i=m-1; i>0; i--){
            table[i]=table[i-1];
        }
        table[0]=-1;
        return table;
    }
}